---
output: github_document
---

<!-- README.md is generated from README.Rmd. Please edit that file -->

```{r, echo = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  fig.path = "README-"
)
```

# powsimR: Power analysis for bulk and single cell RNA-seq experiments

## Installation Guide

You can install powsimR from GitHub using devtools:

```{r, echo=T, eval=F, tidy=T}
library(devtools)
install_github("bvieth/powsimR")
```

If the default installation does not work, you can install powsimR by first installing its dependencies:

```{r, echo=T, eval=F, tidy=T}
ipak <- function(pkg, repository=c('CRAN', 'Bioconductor', 'github')){
  new.pkg <- pkg[!(pkg %in% installed.packages()[, "Package"])]
  if (length(new.pkg)) {
    if(repository=='CRAN') {
      install.packages(new.pkg, dependencies = TRUE)
    }
    if(repository=='Bioconductor') {
      source("https://bioconductor.org/biocLite.R")
      biocLite(new.pkg, dependencies=TRUE, ask=FALSE)
    }
    if(repository=='github') {
      devtools::install_github(new.pkg, build_vignettes = FALSE, dependencies=TRUE)
    }
  }
}

# CRAN PACKAGES
cranpackages <- c("methods", "stats", "matrixStats", "Rtsne", "moments", 
                  "minpack.lm", "glmnet", "cluster", "mclust", "MASS", 
                  "gtools", "doParallel", "parallel", "snow", "reshape2", 
                  "plyr", "dplyr", "tidyr", "tibble", "data.table", "ggplot2", 
                  "grid", "ggthemes", "ggExtra", "cowplot", "scales", 
                  "cobs", "msir", "drc", "DrImpute", "VGAM", "NBPSeq")
ipak(cranpackages, repository='CRAN')

# BIOCONDUCTOR
biocpackages <- c("S4Vectors", "DEDS", "AnnotationDbi", "Biobase", 
                  "BiocGenerics", "SummarizedExperiment", "BiocParallel", 
                  "RUVSeq", "scran", "scater", "SingleCellExperiment", "Linnorm", 
                  "edgeR", "limma", "DESeq2", "baySeq", "NOISeq", "EBSeq", 
                  "MAST", "scde", "scDD", "ROTS", "monocle", "IHW", "qvalue")
ipak(biocpackages, repository='Bioconductor')

# GITHUB
githubpackages <- c('nghiavtr/BPSC', 
                    'rhondabacher/SCnorm', 
                    'catavallejos/BASiCS')
ipak(githubpackages, repository = 'github')

```

After installing the dependencies, powsimR can be installed by using devtools as well.

```{r, echo=T, eval=F, tidy=T}
devtools::install_github('bvieth/powsimR', 
                         build_vignettes = TRUE, 
                         dependencies=FALSE)
```

Some users have experienced issues installing powsimR due to Tex compilation errors. If that is the case, you can leave out building the vignette.

## User Guide

For examples and tips on using the package, please see the vignette PDF 
[here](https://github.com/bvieth/powsimR/tree/master/vignettes/powsimR.pdf) 
or open it in R by typing
```{r, echo=T, eval=F, tidy=T}
browseVignettes('powsimR')
```

## Citation

Please use the following entry for citing powsimR.

```{r, echo=T, eval=F, tidy=T}
citation("powsimR")
```

powsimR is published in [Bioinformatics](https://doi.org/10.1101/117150). 
A preprint paper is also on [bioRxiv](https://doi.org/10.1101/117150). 

## Notes

Please send bug reports and feature requests by opening a new issue on 
[this page](https://github.com/bvieth/powsimR/issues).

Note that the error "maximal number of DLLs reached..." might occur due to the loading of many shared objects by Bioconductor packages. Restarting the R session after installing dependencies / powsimR will help.

Starting with R version 3.4.0, one can set the environmental variable 'R_MAX_NUM_DLLS' to a higher number. See `?Startup()` or the [vignette](https://github.com/bvieth/powsimR/tree/master/vignettes/powsimR.pdf) for more information.

## R Session Info

```{r}
sessionInfo()
```


