---
output: github_document
---

<!-- README.md is generated from README.Rmd. Please edit that file -->

```{r, echo = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  fig.path = "README-"
)
```

# powsimR: Power analysis for bulk and single cell RNA-seq experiments

## Installation Guide

To install powsimR, make sure you have installed the following R packages:

```{r, echo=T, eval=F, tidy=T}
ipak <- function(pkg, repository=c("CRAN", "Bioconductor", "github")){
  new.pkg <- pkg[!(pkg %in% installed.packages()[, "Package"])]
  if (length(new.pkg)) {
    if(repository=="CRAN") {
      install.packages(new.pkg, dependencies = TRUE)
    }
    if(repository=="Bioconductor") {
      source("https://bioconductor.org/biocLite.R")
      biocLite(new.pkg, dependencies=TRUE, ask=FALSE)
    }
    if(repository=="github") {
      devtools::install_github(pkg, build_vignettes = FALSE)
    }
  }
}

# CRAN PACKAGES
cranpackages <- c("methods", "stats", "matrixStats", "moments", "minpack.lm", "glmnet", "cluster", "mclust", "MASS", "doParallel", "parallel", "snow", "reshape2", "plyr", "dplyr", "tidyr", "tibble", "data.table", "ggplot2", "ggthemes", "ggExtra", "cowplot", "scales", "fitdistrplus", "MASS", "pscl", "nonnest2", "cobs", "msir", "drc", "devtools", "XML", "splines", "gtools", "NBPSeq", "permute", "DrImpute", "VGAM")
ipak(cranpackages, repository="CRAN")

# BIOCONDUCTOR
biocpackages <- c("S4Vectors", "AnnotationDbi", "Biobase", "BiocParallel", "BiocGenerics", "BiocStyle", "SummarizedExperiment", "scater", "scran", "edgeR", "limma", "DESeq2", "baySeq", "NOISeq", "EBSeq", "DSS", "MAST", "ROTS", "IHW", "qvalue", "scDD", "monocle", "RUVSeq", "DEDS")
ipak(biocpackages, repository="Bioconductor")

# GITHUB
githubpackages <- c("hms-dbmi/scde", "nghiavtr/BPSC", "rhondabacher/SCnorm", "catavallejos/BASiCS")
ipak(githubpackages, repository = "github")

```

After installing the dependencies, powsimR can be installed by using devtools as well.

```{r, echo=T, eval=F, tidy=T}
devtools::install_github("bvieth/powsimR", build_vignettes = TRUE)
```

## Notes

Please send bug reports and feature requests by opening a new issue on 
[this page](https://github.com/bvieth/powsimR/issues).

Note that the error "maximal number of DLLs reached..." might occur due to the loading of many shared objects by Bioconductor packages. Restarting the R session after installing dependencies / powsimR might be enough to handle this.

Starting with R version 3.4.0, one can set the environmental variable "R_MAX_NUM_DLLS" to a higher number so that the error does not occur anymore. See `?Startup()` for more information. Add the line "R_MAX_NUM_DLLS=150" to your Renviron file located in 

## User Guide

For examples and tips on using the package, please see the vignette PDF 
[here](https://github.com/bvieth/powsimR/tree/master/vignettes/powsimR.pdf) 
or open it in R by typing
```{r, echo=T, eval=F, tidy=T}
browseVignettes("powsimR")
```

## Citation

Please use the following entry for citing powsimR.

```{r, echo=T, eval=F, tidy=T}
citation("powsimR")
```

powsimR is published in [Bioinformatics](https://doi.org/10.1101/117150). 
A preprint paper describing powsimR is also on [bioRxiv](https://doi.org/10.1101/117150). 

## R Session Info

```{r}
sessionInfo()
```


