---
output: github_document
---

<!-- README.md is generated from README.Rmd. Please edit that file -->

```{r, echo = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  fig.path = "README-"
)
```

# powsimR: Power analysis for bulk and single cell RNA-seq experiments

## NEWS

**Version 1.1.1** released with the following changes / additions (2018-04-19):

* `simulateDE` now with the option to perform DE testing on filtered/imputed counts (option `DEFilter`)

**Version 1.1.0** released with the following changes / additions (2018-03-29):

* simulation of batch effects (see options `p.B`, `bLFC` and `bPattern` in `DESetup` and `simulateCounts`)
* simulation of spike-in expression (see `estimateSpike` , `plotSpike` and option `spikeIns` in `simulateDE` and `simulateCounts`)
* simulation of multiple sample groups (e.g. single cell populations) with `simulateCounts`
* imputation and prefiltering options prior to normalisation in DE power simulations added (scImpute, scone, Seurat, DrImpute, SAVER)
* additional normalisation options and DE tools (esp. for single cells) included in `simulateDE`
* evaluation of simulation setup using estimated versus true value comparisons of library size factors and log2 fold changes in `evaluateSim` and `plotEvalSim`
* increased flexibility in preprocessing for distribution evaluation in `evaluateDist`

## Installation Guide

For the installation, the R package `devtools` is needed.

```{r, echo=T, eval=F, tidy=T}
install.packages('devtools')
library(devtools)
```

I recommend to install first the dependencies manually and then powsimR: 

```{r, echo=T, eval=F, tidy=T}
ipak <- function(pkg, repository=c('CRAN', 'Bioconductor', 'github')){
  new.pkg <- pkg[!(pkg %in% installed.packages()[, "Package"])]
  # new.pkg <- pkg
  if (length(new.pkg)) {
    if(repository=='CRAN') {
      install.packages(new.pkg, dependencies = TRUE)
    }
    if(repository=='Bioconductor') {
      source("https://bioconductor.org/biocLite.R")
      biocLite(new.pkg, dependencies=TRUE, ask=FALSE)
    }
    if(repository=='github') {
      devtools::install_github(new.pkg, build_vignettes = FALSE, dependencies=TRUE)
    }
  }
}

# CRAN PACKAGES
cranpackages <- c("bbmle", "broom", "cluster", "cobs", "cowplot", 
                  "data.table", "doParallel", "dplyr", "drc", "DrImpute", "fastICA", "fitdistrplus",
                  "foreach", "gamlss.dist", "ggExtra", "ggplot2", "ggthemes", "grDevices",
                  "glmnet", "grid", "gtools", "Hmisc", "kernlab", "MASS", 
                  "matrixStats", "mclust", "methods", "minpack.lm", "moments", "msir", 
                  "NBPSeq", "nonnest2", "parallel", "penalized", "plyr", "pscl",
                  "reshape2", "ROCR", "Rtsne", "scales", "Seurat", "snow", 
                  "stats", "tibble", "tidyr", "VGAM", "ZIM")
ipak(cranpackages, repository='CRAN')

# BIOCONDUCTOR
biocpackages <- c("AnnotationDbi", "baySeq", "Biobase", "BiocGenerics", 
                  "BiocParallel", "DEDS", "DESeq2", "EBSeq", "edgeR", "IHW", "iCOBRA",
                  "limma", "Linnorm", "MAST", "monocle", "NOISeq", "qvalue", "ROTS", "RUVSeq", 
                  "S4Vectors", "scater", "scDD", "scde", "scone", "scran", "SCnorm", 
                  "SingleCellExperiment", "SummarizedExperiment", "zinbwave")
ipak(biocpackages, repository='Bioconductor')

# GITHUB
githubpackages <- c('nghiavtr/BPSC', 'VCCRI/cidr', 'cz-ye/DECENT', 
                    'mohuangx/SAVER', 'statOmics/zingeR')
ipak(githubpackages, repository = 'github')

```

After installing the dependencies, powsimR can be installed by using devtools as well.

```{r, echo=T, eval=F, tidy=T}
devtools::install_github('bvieth/powsimR', 
                         build_vignettes = TRUE, 
                         dependencies=FALSE)
library("powsimR")
```

Alternative, you can try to install powsimR and its dependencies directly using devtools:

```{r, echo=T, eval=F, tidy=T}
devtools::install_github("bvieth/powsimR")
```

Some users have experienced issues installing powsimR due to vignette compilation errors. If that is the case, you can leave out building the vignette.

### DLLs and ulimit

Note that the error "maximal number of DLLs reached..." might occur due to the loading of many shared objects by Bioconductor packages. Restarting the R session after installing dependencies / powsimR will help.
Starting with R version 3.4.0, one can set the environmental variable 'R_MAX_NUM_DLLS' to a higher number. See `?Startup()` for more information. I recommend to increase the maximum number of DLLs that can be loaded to at least 500. The environmental variable R\_MAX\_NUM\_DLLS can be set in R\_HOME/etc/Renviron prior to starting R. For that locate the Renviron file and add the following line: R\_MAX\_NUM\_DLLS=xy where xy is the number of DLLs.
On my Ubuntu machine, the Renviron file is in /usr/lib/R/etc/ and I can set it to 500.

In addition, the user limits for open files (unix: ulimit) might have to be set to a higher number to accomodate the increase in DLLs. Please check out the help pages for [MACs](https://gist.github.com/tombigel/d503800a282fcadbee14b537735d202c) and [Linux](https://glassonionblog.wordpress.com/2013/01/27/increase-ulimit-and-file-descriptors-limit/) for guidance.

## User Guide

For examples and tips on using the package, please see the vignette 
[here](https://github.com/bvieth/powsimR/tree/master/vignettes/powsimR.html) 

## Citation

Please use the following entry for citing powsimR.

```{r, echo=T, eval=F, tidy=T}
citation("powsimR")
```

powsimR is published in [Bioinformatics](https://doi.org/10.1093/bioinformatics/btx435). 
A preprint paper is also on [bioRxiv](https://doi.org/10.1101/117150). 

## Notes

Please send bug reports and feature requests by opening a new issue on 
[this page](https://github.com/bvieth/powsimR/issues).



## R Session Info

```{r}
library(powsimR)
sessionInfo()
```


