%\VignetteIndexEntry{Bioconductor LaTeX Style}
%\VignettePackage{BiocStyle}
%\VignetteEngine{knitr::knitr}

\documentclass{article}

<<style, eval=TRUE, echo=FALSE, results='asis'>>=
BiocStyle::latex()
@

\renewcommand{\baselinestretch}{1.25}

\newcommand{\exitem}[3]{%
  \item \texttt{\textbackslash#1\{#2\}} #3 \csname#1\endcsname{#2}.%
}

\bioctitle[powsimR]{powsimR: Power Analysis and Sample Size Estimation for Bulk and Single Cell RNA-Seq Experiments}
\author{Beate Vieth \footnote{vieth@bio.lmu.de}}

\begin{document}

\maketitle

% \packageVersion{\Sexpr{BiocStyle::pkg_ver("powsimR")}}

Report issues on \url{https://github.com/bvieth/powsimR/issues}

\newpage

\tableofcontents

<<setup, eval=TRUE, include=FALSE, cache=FALSE>>=

knitr::opts_chunk$set(xtable.comment = FALSE,
                      echo=F,
                      eval=F,
                      message = F,
                      warning=F,
                      error = F,
                      cache = F,
                      tidy = T,
                      results='hide',
                      fig.lp='fig:',
                      fig.align = 'center',
                      fig.path='vignette-',
                      tidy.opts=list(width.cutoff=60)
                     )

# set the width of your R code output
options(width=60)

@

\newpage

%---------------------------------------------------------
\section{Installation Guide}
%---------------------------------------------------------

PowsimR has a number of dependencies that need to be installed before hand (see also the README file on github). I recommend to increase the maximum number of DLLs that can be loaded to at least 500. The environmental variable R\_MAX\_NUM\_DLLS can be set in R\_HOME/etc/Renviron prior to starting R (for further details see \Rfunction{Startup}). For that locate the Renviron file and add the following line: R\_MAX\_NUM\_DLLS=xy where xy is the number of DLLs.
On my Ubuntu machine, the Renviron file is in /usr/lib/R/etc/ and I can set it to 500. Apparently this is not possible on MACs where the maximum that worked was 153. This is enough for powsimR, however.

<<echo=T, eval=F>>=
ipak <- function(pkg, repository=c('CRAN', 'Bioconductor', 'github')){
  new.pkg <- pkg[!(pkg %in% installed.packages()[, "Package"])]
  if (length(new.pkg)) {
    if(repository=='CRAN') {
      install.packages(new.pkg, dependencies = TRUE)
    }
    if(repository=='Bioconductor') {
      source("https://bioconductor.org/biocLite.R")
      biocLite(new.pkg, dependencies=TRUE, ask=FALSE)
    }
    if(repository=='github') {
      devtools::install_github(new.pkg, build_vignettes = FALSE, dependencies=TRUE)
    }
  }
}

# CRAN PACKAGES
cranpackages <- c("methods", "stats", "matrixStats", "Rtsne", "moments",
                  "minpack.lm", "glmnet", "cluster", "mclust", "MASS",
                  "gtools", "doParallel", "parallel", "snow", "reshape2",
                  "plyr", "dplyr", "tidyr", "tibble", "data.table", "ggplot2",
                  "grid", "ggthemes", "ggExtra", "cowplot", "scales",
                  "cobs", "msir", "drc", "DrImpute", "VGAM", "NBPSeq")
ipak(cranpackages, repository='CRAN')

# BIOCONDUCTOR
biocpackages <- c("S4Vectors", "DEDS", "AnnotationDbi", "Biobase",
                  "BiocGenerics", "SummarizedExperiment", "BiocParallel",
                  "RUVSeq", "scran", "scater", "SingleCellExperiment",
                  "Linnorm", "edgeR", "limma", "DESeq2", "baySeq", "NOISeq", "EBSeq",
                  "MAST", "scde", "scDD", "ROTS", "monocle", "IHW", "qvalue")
ipak(biocpackages, repository='Bioconductor')

# GITHUB
githubpackages <- c('nghiavtr/BPSC',
                    'rhondabacher/SCnorm',
                    'catavallejos/BASiCS')
ipak(githubpackages, repository = 'github')
@

After installing the dependencies, \Rpackage{powsimR} can be installed by using devtools as well.

<<echo=T, eval=F, tidy=T>>=
devtools::install_github('bvieth/powsimR',
                         build_vignettes = TRUE,
                         dependencies=FALSE)

library(powsimR)
@

Some users have experienced issues installing powsimR due to Tex compilation errors. If that is the case, you can leave out building the vignette.

%---------------------------------------------------------
\section{Introduction}
%---------------------------------------------------------

In this vignette, we illustrate the features of \Rpackage{powsimR} by assessing the power to detect differential expression between two groups of embryonic stem cells cultured in standard 2i medium (\href{http://www.ebi.ac.uk/arrayexpress/experiments/E-MTAB-2600/}{E-MTAB-2600}) \cite{Kolodziejczyk2015-vn}.

%---------------------------------------------------------
\section{\Rpackage{powsimR} Workflow}
%---------------------------------------------------------

The basic workflow of \Rpackage{powsimR} is illustrated in figure \ref{fig:powsim_scheme}: A) The mean-dispersion relationship is estimated from RNA-seq data, which can be either single cell or bulk data. The users can provide their own count tables, download from a database (see below) or one of our example data sets (\href{https://github.com/bvieth/powsimRData}{powsimRData}). B) These distribution parameters are then used to set up the simulations. For better comparability, the parameters for the simulation of differential expression are set separately. C) Finally, the TPR and FDR are calculated. Both can be either returned as marginal estimates per sample configuration, or stratified according to the estimates of mean expression, dispersion or dropout rate.

\begin{figure}[h]
\centering
\includegraphics[width=0.99\linewidth]{powsimR_vignette_schematic.pdf}
\caption{\textbf{PowsimR schematic overview.} (A) Estimation: (B) Simulation: (C) Evaluation. Functions are given in orange.}
\label{fig:powsim_scheme}
\end{figure}

\section{Parameter Estimation}

\subsection{Gene Expression}

The parameters of the (zero-inflated) negative binomial distribution, i.e. mean and dispersion are estimated by the function \Rfunction{estimateParam}. In addition, the dropout probability, i.e. the fraction of zero counts per gene, is calculated.
The user can choose between two estimation frameworks:
\begin{description}
\item[NB] Negative binomial distribution.
\item[ZINB] Zero-inflated negative binomial distribution.
\end{description}

In both cases matching moments estimation of mean and dispersion are based on normalized counts.

The user can choose between multiple normalisation methods (see details section of \Rfunction{estimateParam}). Furthermore, a number of methods are group sensitive, e.g. batch labels, namely scran SCnorm, Census, BASiCS, RUV.

The estimates, sequencing depth and normalisation factors are plotted with \Rfunction{plotParam}.

With the following command, we estimate and plot the parameters for the embryonic stem cells cultured in standard 2i+lif medium (Kolodziejczyk) (figure \ref{fig:NBparams}). As expected for single cell RNA-seq, the variability (i.e. dispersion) and dropout rates are high. Furthermore, the dispersion strongly depends on the mean and does not level off with higher mean values.

<<echo=TRUE,eval=FALSE>>=
# download count table
githubURL <- "https://github.com/bvieth/powsimRData/raw/master/data-raw/kolodziejczk_cnts.rda"
download.file(url = githubURL, destfile= "kolodziejczk_cnts.rda", method = "wget")
load('kolodziejczk_cnts.rda')
kolodziejczk_cnts <- kolodziejczk_cnts[, grep('standard', colnames(kolodziejczk_cnts))]
colnames(kolodziejczk_cnts) <- paste0(colnames(kolodziejczk_cnts), "_", 1:ncol(kolodziejczk_cnts))
TwoiLIF.params <- estimateParam(countData=kolodziejczk_cnts,
                                batchData = NULL,
                                spikeData = NULL,
                                spikeInfo = NULL,
                                Lengths = NULL,
                                MeanFragLengths = NULL,
                                Distribution = 'ZINB',
                                RNAseq = 'singlecell',
                                normalisation = 'scran',
                                sigma = 1.96,
                                NCores = NULL)
plotParam(TwoiLIF.params)
@

We have implemented a read count simulation framework assuming either a negative binomial distribution or a zero-inflated negative binomial distribution.
To predict the dispersion given a random draw of mean expression value observed, we apply a locally weighted polynomial regression fit. To capture the variability of dispersion estimates observed, a local variability prediction band is applied.
For bulk RNA-seq experiments, dropouts are less probable but can still occur. To include this phenomenon we sample from the observed dropout rates for genes that have a mean expression value below 5\% dropout probability determined by a decrease constrained B-splines regresssion of dropout rate against mean expression (\Rfunction{cobs} in \CRANpkg{cobs}).
The resulting read count matrix has similar distributional characteristics as the original Kolodziejczyk data set (figure \ref{fig:simeval}).
For the zero-inflated negative binomial distribution, the mean-dispersion relation is similarly estimated, but based on positive read counts. Furthermore, the dropouts are also predicted based on a locally weighted polynomial regression fit between mean and dropouts. Of note, this fit is done separately for amplified and non-amplified transcripts separately and similar proportions of genes as observed are also generated in the simulations \cite{Ziegenhain2017-sf}.
We have found that the negative binomial distribution is particularly suited for UMI-methods (e.g. SCRB-Seq, Drop-Seq, 10XGenomics) \cite{Vieth2017-lh}.

\begin{figure}[h]
\centering
\includegraphics[width=0.95\linewidth]{NBparams.pdf}
\caption{\textbf{Estimated parameters for Kolodziejczyk data set.} A) Sequencing depth per sample with median sequencing depth (grey dashed line). B) Library size normalisation factor per sample with median size factor (grey dashed line). C) Marginal Distribution of log2(mean), log2(dispersion) and dropout. D) Local polynomial regression fit between log2(mean) and log2(dispersion) estimates with variability band (yellow) per gene. Common dispersion estimate (grey dashed line). E) Fraction of dropouts versus estimated mean expression per gene.}
\label{fig:NBparams}
\end{figure}

\begin{figure}[h]
\centering
\includegraphics[width=0.7\linewidth]{simeval.jpeg}
\caption{\textbf{Comparison of estimated and simulated read counts} (A) Dispersion versus Mean. (B) Dropout versus Mean.}
\label{fig:simeval}
\end{figure}

\subsection{Spike-ins}

Some normalisation methods can use spike-ins as part of their normalisation (SCnorm, RUV, Census). To use spike-in information in the simulations, their distributional characteristics need to be estimated. We follow the estimation and simulation framework presented in \cite{Kim2015-zo} where the variance is decomposed into shot noise and mRNA loss due to capture and sequencing efficiency. In short, the parameters for a Beta-distribution describes the RNA molecule capture efficiency and the parameters of a Gamma distribution describes the sequencing efficiency, which we can then use to simulate in silico spike-ins given a mean expression value. We assume that biological variance does not contribute to spike-in expression.
The user needs to define the spike-in expression table and the spike-in information table (IDs, molecules, lengths per spike-in) in the function \Rfunction{estimateSpike}.
The following formula can be used for calculating the number of molecules of spike-ins:

  \begin{displaymath}
    Y_{j} = c_{j} * V * 10^{-3} * D^{-1} * 10^{-18} * {Avogadro}, \quad j=1,\dots,92
  \end{displaymath}

The number of molecules $Y_{j}$ for each ERCC spike-in species is the product of the molar concentration $c_{j}$ (attomoles per microlitre), the dilution factor $1/D$, the volume $V$ (nanolitre), Avogadros' constant ($6.02214129*10^{23}$) and conversion factors between unit scales.

% The molecule calculation is implemented in \Rfunction{calculateSpike}.

%If spike-ins were not added to the original experiment there is the possibility to simulate them in silico. For that, we implemented the framework presented in \cite{Qiu2017-qa}.

%The estimated parameters can be plotted with \Rfunction{plotSpike}.

%---------------------------------------------------------
\section{Simulations}
%---------------------------------------------------------
For simulating differential expression between two groups, the number of genes, number of simulations, percentage of differential expression and effect size are set up with the function \Rfunction{DESetup}. The effect size is here defined as the log2 fold change which can be a constant, sampled from a vector or function. The uniform, normal and gamma distributions are possible options and illustrated in figure \ref{fig:lfcs}. Depending on the settings, these distribution can be broader or narrower. If using this option, we recommend to choose a distribution that closely resembles previously observed or expected fold changes.

The distribution estimates and these settings are then combined to one object with \Rfunction{SimSetup}. This allows the user to assess power of multiple groupwise comparisons and different differential testing methods.
The following command sets up simulations with 10,000 genes, 20\% genes being DE, log fold change sample from a narrow gamma distribution and parameter estimates based on Kolodziejczyk data:

<<echo=TRUE,eval=FALSE>>=

lfc.gamma = function(x) sample(c(-1,1), size=x,replace=T)*rgamma(x, 3, 3)
de.opts = DESetup(ngenes=10000, nsims=25,
                  p.DE=0.2, pLFC=lfc.gamma,
                  p.B = NULL, bLFC = NULL,
                  sim.seed = 58673)
sim.opts = SimSetup(desetup = de.opts,
                    params = TwoiLIF.params,
                    size.factors='given',
                    downsample=FALSE, geneset = FALSE)
@

With the setup defined, the differential expression simulation is run with \Rfunction{simulateDE}. For this, the user needs to set the following options:

\begin{description}
  \item[Replicates] The number of sample replicates per group (n1 and n2). These can be unbalanced.
  \item[DEmethod] The differential testing method. The user can choose between 11 methods in total.  7 developed for bulk, 4 developed for single cells (see the detail section of \Rfunction{simulateDE}).
  \item[normalisation] The normalisation method. The user can choose between 11 methods in total.  6 developed for bulk, 5 developed for single cells (see the detail section of \Rfunction{estimateParam}).
\end{description}

<<echo=TRUE,eval=FALSE>>=
simDE = simulateDE(n1 = c(24,48,96,192,384,800),
                   n2 = c(24,48,96,192,384,800),
                   sim.settings = sim.opts,
                   DEmethod = "MAST",
                   normalisation = "scran",
                   Preclust=FALSE,
                   Preprocess = NULL,
                   spikeIns=FALSE,
                   NCores=10,
                   verbose=TRUE)

@

There also additional options: Whether to apply a preprocessing step prior to normalisation; whether spike-in information should be used (if available).

\begin{figure}[h]
\centering
\includegraphics[width=0.8\linewidth]{lfcdist.jpeg}
\caption{Log2 fold change examples for gamma, uniform and normal distribution}
\label{fig:lfcs}
\end{figure}

% \subsection{Read count matrices}
%
% Some users might be interested in outputting only simulated read counts by calling \Rfunction{simulateCounts}.
%

\section{Evaluation}

The results of differential expression simulation are evaluated with \Rfunction{evaluateDE}. We have separated the evaluation from DE detection to allow the user to evaluate power in a comprehensive way as advocated by \cite{Wu2015-uj}.
In this function, the proportions and error rates are estimated. The rates can be stratified by mean, dispersion dropout or log fold change. Furthermore, the user can choose between different multiple testing correction methods (see \Rfunction{p.adjust.methods}, \Rfunction{ihw} in \Biocpkg{IHW} and \Rfunction{qvalue} in \Biocpkg{qvalue}). Also, the genes can be filtered by mean, dispersion or dropout. To define biologically interesting genes, a cutoff for the log2 fold change with delta can be set.

With the following command we evaluate the marginal TPR and FDR conditional on the mean expression for the simulation based on Kolodziejczyk data.

<<echo=TRUE,eval=FALSE>>=
evalDE = evaluateDE(simRes = simDE,
                     alpha.type = 'adjusted',
                     MTC = 'BH',
                     alpha.nominal = 0.1,
                     stratify.by = 'mean',
                     filter.by = 'none',
                     strata.filtered = 1,
                     target.by = 'lfc',
                     delta = 0)
@

The results of the evaluation can be plotted with \Rfunction{plotEvalDE}.
\begin{description}
  \item[rate] Marginal or conditional error rates calculations. The conditional error rates are determined and calculated with \Rfunction{evaluateDE}. The number of genes per stratum are also summarised.
  \item[quick] If this is set to \R{TRUE} then only the TPR and FDR will be plotted.
\end{description}

With the following commands, the quick marginal and conditional power assessments for the Kolodziejczyk data are plotted.

<<echo=TRUE,eval=FALSE>>=
plotEvalDE(evalRes = evalDE,
            rate='marginal',
            quick=TRUE, annot=TRUE)

plotEvalDE(evalRes = evalDE,
            rate='stratified',
            quick=TRUE, annot=TRUE)
@

% \begin{figure}[h]
% \centering
% \includegraphics[width=0.8\linewidth]{margpower.pdf}
% \caption{Log2 fold change examples for gamma, uniform and normal distribution}
% \label{fig:margpower}
% \end{figure}
%
% \begin{figure}[h]
% \centering
% \includegraphics[width=0.8\linewidth]{condpower.pdf}
% \caption{Log2 fold change examples for gamma, uniform and normal distribution}
% \label{fig:condpower}
% \end{figure}

%---------------------------------------------------------
\section{Additional Functionalities}
%---------------------------------------------------------

\subsection{Evaluate Simulation Framework}

It is important to validate the appropiateness of the chosen simulation framework. The function \Rfunction{evaluateDist} compares the theoretical fit of the Poisson, negative binomial, zero-inflated Poisson and zero-inflated negative binomial and beta-Poisson distribution to the empirical RNA-seq read counts (\cite{Colin_Cameron2013-vb}, \cite{Kim2013-qo}, \cite{Delmans2016-ef}).
The evaluation is then plotted with the function \Rfunction{plotEvalDist} which summarizes the best fitting distribution per gene based on goodness-of-fit statistics (Chi-square test), Akaike Information Criterium, comparing observed dropouts with zero count prediction of the models and comparing the model fitness with Likelihood Ratio Test and Vuong Test.
% As noted by other developers, goodness-of-fit tests are not an objective tool and heavily depend on sample sizes (\cite{Delignette-Muller2015-ie}). A graphical evaluation of the fitted distribution is considered the most appropiate way but for high-throughput sequencing an unrealistic recommendation.
Bulk RNA-seq experiments are usually conducted with a small number of samples. We therefore recommend to rely on the goodness-of-fit validation by \cite{Mi2015-ri}. To use this approach in \Rfunction{evaluateDist}, the user should allow for permutation simulations by setting the value of nsims to at least 100. If available, the computation can be run on multiple cores by setting the number of cores (ncores).

With the following command, we estimate and plot the parameters for the embryonic stem cells cultured in standard 2i lif medium (Kolodziejczyk).

<<echo=TRUE,eval=FALSE>>=
TwoiLIF.distfit = evaluateDist(cnts = kolodziejczk_cnts,
                             RNAseq = 'singlecell',
                             ncores = 1,
                             nsims = 1,
                             frac.genes = 1,
                             min.meancount = 1,
                             min.libsize =1000)

plotEvalDist(evalDist = TwoiLIF.distfit, annot=F)
@

\begin{figure}[h]
\centering
\includegraphics[width=0.75\linewidth]{evaldist.pdf}
\caption{A) Goodness of fit of the model per gene assessed with a Chi-square test based on residual deviance and degrees of freedom. B) The fraction of genes for which the respective distribution has the lowest AIC and additionally the distribution with the lowest AIC as well as not rejected by the goodness of fit statistic.  C) Observed versus predicted dropouts per distributional model and gene. D) Model assessment per gene based on Likelihood Ratio Test for nested models and Vung Test for non-nested models.}
\label{fig:evaldist}
\end{figure}

\subsection{Negative Binomial Parameters}

\subsubsection{in silico Parameter Definition}

We have also implemented the option to approximate the read count matrix simulation based on random distribution functions in \R{}. The user then has to define the mean, dispersion, dropout and library size in \Rfunction{insilicoNBParam}. In the absence of suitable pilot studies, a typical single cell RNA-seq experiment could be approximated with:
\begin{itemize}
  \item mean: \Rcode{function(x) rgamma(x, 4, 2)} where x is the number of genes
  \item dispersion: \Rcode{function(x) 2 + 100/x} where x is the mean
  \item library size: \Rcode{function(x) 2**rnorm(n=x, mean=0, sd=0.25)} where x is the number of samples
\end{itemize}

The same functionality can also be used for bulk RNA-seq.

\subsubsection{Count matrices of single cell RNA-seq experiments}

We have uploaded read count matrices of 5 single cell RNA-seq experiments on \href{https://github.com/bvieth/powsimRData}{github}.
The user can calculate the negative binomial parameters with \Rfunction{estimateParam}, view these estimates with \Rfunction{plotParam} and use it as an input for \Rfunction{SimSetup}.

\subsubsection{Access to raw read counts stored in online data base}

We have provided a number of exemplatory single cell RNA-seq data sets for parameter estimation. Nevertheless, you might not find a data set that matches your own experimental setup. In those cases, we recommend to check online repositories for a suitable data set. Below you can find an example script to get count tables from recount (\url{https://jhubiostatistics.shinyapps.io/recount/}) \cite{Collado-Torres2017-mo}. For a single cell RNA-seq data base, see conquer and its tutorial (\url{http://imlspenticton.uzh.ch:3838/conquer/}).
As before, the user can then calculate the negative binomial parameters with \Rfunction{estimateParam}, view these estimates with \Rfunction{plotParam} and use it as an input for \Rfunction{SimSetup}.

<<online_repos, echo=T, eval=F>>=
# Install and load the R package
source('http://bioconductor.org/biocLite.R')
biocLite('recount')
library('recount')

# Download the data set
url <- download_study('SRP060416')

# Load the data
load(file.path('SRP060416', 'rse_gene.Rdata'))

# count table
cnts <- assay(rse_gene)
# sample annotation
sample.info <- data.frame(colData(rse_gene)@listData, stringsAsFactors=F)
# gene annotation
gene.info <- data.frame(GeneID=rowData(rse_gene)@listData$gene_id, GeneLength=rowData(rse_gene)@listData$bp_length, stringsAsFactors=F)
@


\subsection{Simulation settings}

By default, there is no difference in library sizes between the samples. If the user wishes for a more realistic, i.e. more variable distribution of read counts across samples, the library sizes can be sampled from observed, vector or function.

% %---------------------------------------------------------
% \section{Wrapper Function}
% %---------------------------------------------------------
%
% \Rfunction{PowSim} is a wrapper including estimation, simulation and evaluation. Please consult the detailed description of \Rfunction{PowSim} help page for more information.
% <<echo=T, eval=F>>=
% res <- PowSim(input=NULL, RNAseq='singlecell', ngenes=10000, nsims=25, p.DE=0.1, LFC=function(x) sample(c(-1,1), size=x,replace=T)*rgamma(x, 3, 3), size.factors='equal', ncores=10, DEmethod="MAST", save.plots=TRUE, verbose=TRUE)
% @


%---------------------------------------------------------
\section{Session info}
%---------------------------------------------------------

Here is the output of \Rfunction{sessionInfo} on the system on which
this document was compiled:
<<sessionInfo, results='asis', echo=FALSE, eval=TRUE>>=
toLatex(sessionInfo())
@

\bibliography{Bioc}

\end{document}
