<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html>
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Simulate Single Cell Analysis Pipeline Expression — simulateFlow • powsimR</title>

<!-- jquery -->
<script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css" rel="stylesheet" integrity="sha384-BVYiiSIFeK1dGmJRAkycuHAHRg32OmUcww7on3RYdg4Va+PmSTsz/K68vbdEjh4u" crossorigin="anonymous">
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/1.7.1/clipboard.min.js" integrity="sha384-cV+rhyOuRHc9Ub/91rihWcGmMmCXDeksTtCihMupQHSsi8GIIRDG0ThDc3HGQFJ3" crossorigin="anonymous"></script>

<!-- sticky kit -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>



<meta property="og:title" content="Simulate Single Cell Analysis Pipeline Expression — simulateFlow" />

<meta property="og:description" content="Blabla" />
<meta name="twitter:card" content="summary" />



<!-- mathjax -->
<script src='https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML'></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->


  </head>

  <body>
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">powsimR</a>
        <span class="label label-default" data-toggle="tooltip" data-placement="bottom" title="Released package">1.1.2</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/powsimR.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul>
      
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/bvieth/powsimR">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      
      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Simulate Single Cell Analysis Pipeline Expression</h1>
    <small class="dont-index">Source: <a href='https://github.com/bvieth/powsimR/blob/master/R/Flow.R'><code>R/Flow.R</code></a></small>
    <div class="hidden name"><code>simulateFlow.Rd</code></div>
    </div>

    <div class="ref-description">
    
    <p>Blabla</p>
    
    </div>

    <pre class="usage"><span class='fu'>simulateFlow</span>(<span class='kw'>n1</span><span class='kw'>=</span><span class='fu'>c</span>(<span class='fl'>20</span>,<span class='fl'>50</span>,<span class='fl'>100</span>), <span class='kw'>n2</span><span class='kw'>=</span><span class='fu'>c</span>(<span class='fl'>30</span>,<span class='fl'>60</span>,<span class='fl'>120</span>),
<span class='no'>sim.settings</span>,
<span class='no'>DEmethod</span>,
<span class='no'>normalisation</span>,
<span class='kw'>Preclust</span><span class='kw'>=</span><span class='fl'>FALSE</span>,
<span class='kw'>Prefilter</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
<span class='kw'>Impute</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
<span class='kw'>DEFilter</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>,
<span class='kw'>GeneSelect</span><span class='kw'>=</span><span class='kw'>NULL</span>,
<span class='kw'>DimReduce</span><span class='kw'>=</span><span class='kw'>NULL</span>,
<span class='kw'>ClustMethod</span><span class='kw'>=</span><span class='kw'>NULL</span>,
<span class='kw'>Pipeline</span><span class='kw'>=</span><span class='kw'>NULL</span>,
<span class='kw'>spikeIns</span><span class='kw'>=</span><span class='fl'>FALSE</span>,
<span class='kw'>NCores</span><span class='kw'>=</span><span class='kw'>NULL</span>,
<span class='kw'>verbose</span><span class='kw'>=</span><span class='fl'>TRUE</span>)</pre>
    
    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>n1, n2</th>
      <td><p>Integer vectors specifying the number of biological replicates in each group. Default values are n1=c(20,50,100) and n2=c(30,60,120).</p></td>
    </tr>
    <tr>
      <th>sim.settings</th>
      <td><p>This object specifies the simulation setup. This must be the return object from <code><a href='SimSetup.html'>SimSetup</a></code>.</p></td>
    </tr>
    <tr>
      <th>DEmethod</th>
      <td><p>A character vector specifying the DE detection method to be used.
Please consult the Details section for available options.</p></td>
    </tr>
    <tr>
      <th>normalisation</th>
      <td><p>Normalisation method to use.
Please consult the Details section for available options.</p></td>
    </tr>
    <tr>
      <th>Preclust</th>
      <td><p>A logical vector indicating whether to run a hierarchical clustering prior to normalisation.
This is implemented for scran only. Default is <code>FALSE</code>.
For details, see <code><a href='http://www.rdocumentation.org/packages/scran/topics/quickCluster'>quickCluster</a></code>.</p></td>
    </tr>
    <tr>
      <th>Prefilter</th>
      <td><p>A character vector specifying the gene expression filtering method
to be used prior to normalisation (and possibly imputation).
Default is <code>NULL</code>, i.e. no filtering.
Please consult the Details section for available options.</p></td>
    </tr>
    <tr>
      <th>Impute</th>
      <td><p>A character vector specifying the gene expression imputation method
to be used prior to normalisation.
Default is <code>NULL</code>, i.e. no imputation.
Please consult the Details section for available options.</p></td>
    </tr>
    <tr>
      <th>DEFilter</th>
      <td><p>A logical vector indicating whether to run DE testing on filtered and/or imputed count data.
Default is <code>FALSE</code>.</p></td>
    </tr>
    <tr>
      <th>GeneSelect</th>
      <td><p>A character vector specifying the gene selection method.
Default is <code>NULL</code>, i.e. no gene selection.
Availabe options are: HVG, Gini-Index.</p></td>
    </tr>
    <tr>
      <th>DimReduce</th>
      <td><p>A character vector specifying the dimension reduction method. Default is <code>NULL</code>, i.e. no dimension reduction.
Availabe options are: Euclidean, Spearman, PCA, t-SNE-expr, t-SNE-Euclidean, PCA+t-SNE, CIDR.</p></td>
    </tr>
    <tr>
      <th>ClustMethod</th>
      <td><p>A character vector specifying the clustering method. Default is <code>NULL</code>, i.e. no clustering.
Availabe options are: kmeans, PAM.</p></td>
    </tr>
    <tr>
      <th>Pipeline</th>
      <td><p>A character vector specifying out-of-the-box single cell clustering pipelines.
Available options are: CIDR_free and CIDR_bound.</p></td>
    </tr>
    <tr>
      <th>spikeIns</th>
      <td><p>Logical value to indicate whether to simulate spike-ins. Default is <code>FALSE</code>.</p></td>
    </tr>
    <tr>
      <th>NCores</th>
      <td><p>integer positive number of cores for parallel processing, default is <code>NULL</code>, ie 1 core.</p></td>
    </tr>
    <tr>
      <th>verbose</th>
      <td><p>Logical value to indicate whether to show progress report of simulations. Default is <code>TRUE</code>.</p></td>
    </tr>
    </table>
    
    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>A list with the following fields:</p>
<dt>pvalue, fdr</dt><dd><p>3D array (ngenes * N * nsims) for p-values and FDR from each simulation.
Note that FDR values will be empty and the calculation will be done by <code><a href='evaluateDE.html'>evaluateDE</a></code> whenever applicable.</p></dd>
<dt>mu,disp,dropout</dt><dd><p>3D (ngenes * N * nsims) array for mean, dispersion and dropout of library size factor normalized read counts.</p></dd>
<dt>elfc,rlfc</dt><dd><p>3D array (ngenes * N * nsims) for log2 fold changes (LFC):
elfc is for the DE tool estimated LFC; rlfc is for the LFC estimated from the normalised read counts.</p></dd>
<dt>sf.values,gsf.values</dt><dd><p>3D array (ngenes * N * nsims) for size factor estimates.
Global estimates per sample in sf.values; Gene- and sample-wise estimates in gsf.values only for SCnorm normalisation.</p></dd>
<dt>true.designs,def.designs</dt><dd><p>3D array (ngenes * N * nsims) for group assignment specifications.
true.designs for the simulated group assignment; def.designs for the group assignment determined after clustering.</p></dd>
<dt>sim.settings</dt><dd><p>The input sim.settings to which the specifications of <code>simulateDE</code> is added.</p></dd>
<dt>time.taken</dt><dd><p>The time taken for each simulation, given for preprocessing, normalisation, clustering, differential expression testing and moment estimation.</p></dd>

    
    <h2 class="hasAnchor" id="see-also"><a class="anchor" href="#see-also"></a>See also</h2>

    <div class='dont-index'><p><code><a href='estimateParam.html'>estimateParam</a></code>,  <code><a href='insilicoNBParam.html'>insilicoNBParam</a></code> for negative binomial parameter specifications;<br />
 <code><a href='DESetup.html'>DESetup</a></code>, <code><a href='SimSetup.html'>SimSetup</a></code> for simulation setup</p></div>
    

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"># NOT RUN {
## not yet
# }
</pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      <li><a href="#arguments">Arguments</a></li>
      
      <li><a href="#value">Value</a></li>

      <li><a href="#see-also">See also</a></li>
      
      <li><a href="#examples">Examples</a></li>
    </ul>

    <h2>Author</h2>
    
Beate Vieth

  </div>
</div>

      <footer>
      <div class="copyright">
  <p>Developed by .</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://pkgdown.r-lib.org/">pkgdown</a>.</p>
</div>

      </footer>
   </div>

  

  </body>
</html>

